Mathlib/Data/Finsupp/Fin.lean

```diff
@@ -23,27 +27,124 @@ noncomputable section
 
 namespace Finsupp
 
-variable {n : ℕ} (i : Fin n) {M : Type*} [Zero M] (y : M) (t : Fin (n + 1) →₀ M) (s : Fin n →₀ M)
+variable {n : ℕ} {M : Type*} [Zero M]
 
 /-- `tail` for maps `Fin (n + 1) →₀ M`. See `Fin.tail` for more details. -/
 def tail (s : Fin (n + 1) →₀ M) : Fin n →₀ M :=
   Finsupp.equivFunOnFinite.symm (Fin.tail s)
 
+/-- `init` for maps `Fin n →₀ M`. See `Fin.init` for more details. -/
+def init (s : Fin (n + 1) →₀ M) : Fin n →₀ M :=
+  Finsupp.equivFunOnFinite.symm (Fin.init s)
+
+/-- `removeNth` for maps `Fin n →₀ M`. See `Fin.removeNth` for more details. -/
+def removeNth (p : Fin (n + 1)) (s : Fin (n + 1) →₀ M) : Fin n →₀ M :=
+  Finsupp.equivFunOnFinite.symm (Fin.removeNth p s : Fin n → M)
+
 /-- `cons` for maps `Fin n →₀ M`. See `Fin.cons` for more details. -/
 def cons (y : M) (s : Fin n →₀ M) : Fin (n + 1) →₀ M :=
   Finsupp.equivFunOnFinite.symm (Fin.cons y s : Fin (n + 1) → M)
 
-theorem tail_apply : tail t i = t i.succ :=
-  rfl
+/-- `snoc` for maps `Fin n →₀ M`. See `Fin.snoc` for more details. -/
+def snoc (s : Fin n →₀ M) (y : M) : Fin (n + 1) →₀ M :=
+  Finsupp.equivFunOnFinite.symm (Fin.snoc s y : Fin (n + 1) → M)
+
+/-- `insertNth` for maps `Fin n →₀ M`. See `Fin.insertNth` for more details. -/
+def insertNth (p : Fin (n + 1)) (y : M) (s : Fin n →₀ M) : Fin (n + 1) →₀ M :=
+  Finsupp.equivFunOnFinite.symm (Fin.insertNth p y s : Fin (n + 1) → M)
+
+section Remove
+
+variable (p : Fin (n + 1)) (s : Fin (n + 1) →₀ M) (y : M) (i : Fin n)
+
+@[simp]
+theorem removeNth_apply : removeNth p s i = s (p.succAbove i) := rfl
+
+@[simp]
+theorem removeNth_zero : removeNth 0 s = tail s := rfl
+
+@[simp]
+theorem removeNth_last : removeNth (Fin.last n) s = init s := by
+  ext
+  simp only [removeNth, Fin.removeNth_last, equivFunOnFinite_symm_apply_toFun, init]
+
+/-- Updating the `i`-th entry does not affect removing the `i`-th entry. -/
+@[simp]
+theorem removeNth_update : removeNth p (update s p y) = removeNth p s := by
+  simp only [removeNth, update, coe_mk, EmbeddingLike.apply_eq_iff_eq]
+  convert Fin.removeNth_update p y (⇑s)
+
+@[simp]
+theorem tail_apply : tail s i = s i.succ := rfl
+
+@[simp]
+theorem tail_update_zero : tail (update s 0 y) = tail s := by
+  simp only [tail, update, coe_mk, EmbeddingLike.apply_eq_iff_eq]
+  convert Fin.tail_update_zero (⇑s) y
+
+@[simp]
+theorem tail_update_succ : tail (update s i.succ y) = update (tail s) i y := by
+  ext a
+  have : ⇑(equivFunOnFinite.symm (Fin.tail ⇑s)) = Fin.tail ⇑s := rfl
+  simp only [tail, update, this, coe_mk, equivFunOnFinite_symm_apply_toFun]
+  have : Fin.tail (Function.update (⇑s) i.succ y) a = Function.update (Fin.tail ⇑s) i y a := by
+    simp only [Fin.tail_update_succ]
+  convert this
```

Comment:
Sorry for the delay.

@YaelDillies: the only new lemmas that use the existing definitions (`Finsupp.{tail/cons}`) are these two, which are about the relationship between `tail` and `update`. The rest are dependent on the new definitions (`Finsupp.{init/snoc/removeNth/insertNth}`).

Because of this, I'm not very sure if a dependent PR with two lemmas is of much use (though I'd be glad to do it if it helps).

Reply:
I think it's worth splitting anyway. Thanks 🙏